example/canny.cpp

```diff
@@ -0,0 +1,165 @@
+    auto hor_it = view(horizontal).begin();
+    auto slope_it = slope.begin();
+
+    for (auto mag_it = magnitude.begin(); mag_it != magnitude.end();
+            ++mag_it, ++ver_it, ++hor_it, ++slope_it) {
+
+        *mag_it = std::min(255,
+                           (int)std::sqrt(
+                               std::pow((int)(*ver_it), 2) + std::pow((int)(*hor_it), 2)));
+
+        *slope_it = (float)std::atan2((int)(*hor_it), (int)(*ver_it));
+    }
+}
+
+void suppression_threshold(const gray8c_view_t &magnitude, const gray32fc_view_t &angle,
+                           int minVal, int maxVal, const gray8_view_t &dst)
+{
+    fill_pixels(dst, int8_t(0));
+
+    auto mag_loc = magnitude.xy_at(1, 1);
+    int q, r;
+    int index;
+    int curVal;
+    for (int y = 1; y < magnitude.height() - 1; ++y) {
+        auto dst_it = dst.row_begin(y);
+        auto angle_it = angle.row_begin(y);
+
+        for (int x = 1; x < dst.width() - 1; ++x, ++angle_it, ++dst_it, ++mag_loc.x()) {
+            index = (int)((*angle_it)[0] * 8 / M_PI);
+            index = (index < 0) ? index + 8 : index;
+
+            curVal = mag_loc(0, 0);
+
+            switch (index) {
+            case 0:
+            case 7:
+            case 8:
+                q = mag_loc(1, 0);
+                r = mag_loc(-1, 0);
+                break;
+            case 1:
+            case 2:
+                q = mag_loc(1, 1);
+                r = mag_loc(-1, -1);
+                break;
+            case 3:
+            case 4:
+                q = mag_loc(0, 1);
+                r = mag_loc(0, -1);
+                break;
+            case 5:
+            case 6:
+                q = mag_loc(1, -1);
+                r = mag_loc(-1, 1);
+                break;
+            default:
+                q = 255;
+                r = 255;
+            }
+
+            if ((curVal >= minVal) && (curVal >= q) && (curVal >= r)) {
+                if (curVal >= maxVal)
+                    *dst_it = 255;
+                else
+                    *dst_it = 150;
+            }
+        }
+        mag_loc += point2<std::ptrdiff_t>(-dst.width() + 2, 1);
+    }
+}
+
+void hysteresis(const gray8_view_t &dst)
+{
+    auto dst_loc = dst.xy_at(1, 1);
+    for (int i = 1; i < dst.height() - 1; ++i) {
+        for (int j = 1; j < dst.width() - 1; ++j, ++dst_loc.x()) {
+            if (dst_loc(0, 0) == 150) {
+                if (dst_loc(-1, -1) == 255 || dst_loc(-1, 0) == 255 ||
+                        dst_loc(-1, 1) == 255 || dst_loc(1, -1) == 255 ||
+                        dst_loc(1, 0) == 255 || dst_loc(1, 1) == 255 ||
+                        dst_loc(0, -1) == 255 || dst_loc(0, 1) == 255)
+                    dst_loc(0, 0) = 255;
+                else
+                    dst_loc(0, 0) = 0;
+            }
+        }
+        dst_loc += point2<std::ptrdiff_t>(-dst.width() + 2, 1);
+    }
+}
+
+void canny(const rgb8c_view_t &src, const gray8_view_t &dst, int minVal, int maxVal)
+{
```

Comment:
It would be helpful if there is comment about steps performed by `canny`, i.e. steps of the algorithm.
Alternatively, rename functions to make clearer indications of what they do
or add link to reference paper the implementation follows.